Updating documentation for Topology Aware Routing in 1.27 (#40069)
kubernetes/website: content/en/docs/concepts/services-networking/topology-aware-routing.md

Reviewer: If changing the URL, also update `static/_redirects`

Author: Thanks for the heads up, added to redirects file.

Reviewer: Optional, but recommended: add a 302 (**not** 301) redirect from https://kubernetes.io/docs/concepts/services-networking/topology-aware-hints/ to https://kubernetes.io/docs/concepts/services-networking/topology-aware-routing/

They're configured in `static/_redirects`.

I'm suggesting a 302 because it's not a _move_, it's an equivalent concept that's almost the same.

Reviewer: D'oh, I missed this.

Reviewer: We might want to clarify that https://k8s.io/docs/concepts/services-networking/service-topology/ is a different thing with the same name, and update https://k8s.io/docs/concepts/services-networking/service-topology/ to replace the mention of “Topology aware hints”.

Author: I updated the link from the old page to the new page, mentioned in another concept that I think the best solution may be to remove the old page since it's been deprecated for >2 years now.

---
reviewers:
- robscott

> title: Topology Aware Routing

Reviewer: How will we avoid confusion between this and https://k8s.io/docs/concepts/services-networking/service-topology/?

Author: Can we safely remove the previous doc since it's been removed for 6 k8s versions/2 years and only ever made it to alpha?

{{< feature-state for_k8s_version="v1.23" state="beta" >}}

{{< note >}}

> Prior to Kubernetes 1.27, this feature was known as "Topology Aware Hints".

Author: Updated, thanks!

{{</ note >}}

_Topology Aware Routing_ adjusts routing behavior to prefer keeping traffic in
the zone it originated from. In some cases this can help reduce costs or improve
network performance.

<!-- body -->

## Motivation

Kubernetes clusters are increasingly deployed in multi-zone environments.
_Topology Aware Routing_ provides a mechanism to help keep traffic within the
zone it originated from. When calculating the endpoints for a {{<
glossary_tooltip term_id="Service" >}}, the EndpointSlice controller considers
the topology (region and zone) of each endpoint and populates the hints field to
allocate it to a zone. Cluster components such as {{< glossary_tooltip
term_id="kube-proxy" text="kube-proxy" >}} can then consume those hints, and use
them to influence how the traffic is routed (favoring topologically closer
endpoints).

## Enabling Topology Aware Routing

{{< note >}}
Prior to Kubernetes 1.27, this feature was enabled with the
`service.kubernetes.io/topology-aware-hints` annotation.
{{</ note >}}

> You can enable Topology Aware Hints for a Service by setting the

Reviewer 2: ```suggestion
You can enable Topology Aware Routing for a Service by setting the
```
Should this be part of the name change from Hints to Routing

> You can enable Topology Aware Hints for a Service by setting the

Reviewer 2: ```suggestion
You can enable Topology Aware Routing for a Service by setting the
```
Should this also be changed to Topology Aware Routing?

Author: Updated, thanks!

`service.kubernetes.io/topology-mode` annotation to `Auto`. When there are
enough endpoints available in each zone, Topology Hints will be populated on
EndpointSlices to allocate individual endpoints to specific zones, resulting in
traffic being routed closer to where it originated from.

> ### When It Works Best

Reviewer: (nit)
```suggestion
### When it works best
```

Author: Updated, thanks!

> ### When it works best

Reviewer: Is intended as a subheading of “Enabling Topology Aware Routing?”

Author: Updated, thanks!

This feature works best when:

#### 1. Incoming traffic is evenly distributed

If a large proportion of traffic is originating from a single zone, that traffic
could overload the subset of endpoints that have been allocated to that zone.
This feature is not recommended when incoming traffic is expected to originate
from a single zone.

> #### 1. The Service has 3 or more endpoints per zone (9 endpoints in a 3-zone cluster)

Reviewer: ```suggestion
#### 2. The Service has 3 or more endpoints per zone (9 endpoints in a 3-zone cluster) {#three-or-more-endpoints}
```
Simplifies the fragment identifier

> #### 1. The Service has 3 or more endpoints per zone (9 endpoints in a 3-zone cluster)

Author: Updated, thanks!

> #### 2. The Service has 3 or more endpoints per zone (9 endpoints in a 3-zone cluster) {#three-or-more-endpoints}
> 
> Below this number, there is a high (~50%) probability that the EndpointSlice

Reviewer: Optionally:
```suggestion
#### 2. The Service has 3 or more endpoints per zone {#three-or-more-endpoints-per-zone}

In a three zone cluster, this means 9 or more endpoints. If there are fewer than 3
endpoints per zone, there is a high (≈50%) probability that the EndpointSlice
```

Author: Updated, thanks!

controller will not be able to allocate endpoints evenly and instead will fall
back to the default cluster-wide routing approach.

> ### How It Works

Reviewer: (nit)
```suggestion
## How it works
```

Author: Updated, thanks!

  proportions of each zone. This could have unintended consequences if a large
  portion of nodes are unready.

* The EndpointSlice controller ignores nodes with the

>   `node-role.kubernetes.io/control-plane` or `node-role.kubernetes.io/master`

Reviewer: Optionally, link to the definitions of these labels in https://kubernetes.io/docs/reference/labels-annotations-taints/

Author: Ran out of time for this one, not quite sure how to make this look right and don't want to delay the rest of the PR.

Reviewer 2: Created followup issue for this: https://github.com/kubernetes/website/issues/40602

  either not picking up on this event, or newly added pods starting in a
  different zone.

> ## Custom Heuristics

Reviewer: (nit)
```suggestion
## Custom heuristics
```

> Kubernetes is deployed in many different ways, there is no way that a single

Reviewer 3: ```suggestion
Kubernetes is deployed in many different ways, there is no single
```

Author: Updated, thanks!

> heuristic for allocating endpoints to zones will work for every use case. A

Reviewer 3: ```suggestion
heuristic for allocating endpoints to zones will work for every use cases. A
```

Author: I don't think this is right, at least it doesn't sound quite right to me. I may be missing something though.

feature is to enable custom heuristics to be developed if the built in heuristic

> does not work for your use case. The first steps to enable custom heuristics
> were included in the 1.27 release, but additional updates will be required to
> make this more broadly approachable.

Reviewer: ```suggestion
does not work for your use case. The first steps to enable custom heuristics
were included in the 1.27 release. This is a limited implementation that may not
yet cover some relevant and plausible situations.
```

Author: Updated, thanks!
